### ISSUE TITLE
addition to hs.timer

### ISSUE THREAD

**Issue by asmagill** [Hammerspoon/hammerspoon#463]
I find myself more and more often writing code involving hs.timer which essentially runs once a second checking until some condition is true, then doing something and killing off the timer after performing whatever the action is... basically a "defer doing X until condition Y is met"

Functionally this is really a "repeat until function returns true", so I'm wondering which makes more sense as the name of a new constuctor:  `hs.timer.deferAction(fn, checkInterval)` (which is how I use it) or `hs.timer.repeatWhileTrue(fn, checkInterval)` (which is what it really is)

Basically this function (by whatever name) does the following:
1) create and start a new timer, which repeats every `checkInterval` seconds (1 by default)
2) executes the function
3) if `fn` returns false, stop the timer and delete the locally stored reference to it

(corrected to stop on false -- typo in original)

Thoughts?  Is this even worth adding to the module?

**Comment by asmagill** [comment]
Maybe hs.timer.repeatUntilTrue makes more sense?

**Comment by lowne** [comment]
Another option: from `.doAfter(time,fn)`, `.doWhen(predicatefn,fn[,checkInterval])` (the reference is returned so we can still `:cancel` the timer if we change our mind) and its converse `.doUntil(predicatefn,fn,repeatInterval)`
(furthermore, I'd just simplify them all to .after, .when, .until)

**Comment by asmagill** [comment]
Separate the trigger logic from the (possible) action function... I actually like that better.

Thanks!

**Comment by asmagill** [comment]
"until" is a reserved keyword in Lua and can't be a variable/function name (unless you want to invoke it as hs.timer["until"], but that seems worse to me...)

So, stick with the "do" before all three or come up with another name for until?  I'm ok with the "do" as it re-iterates that we're "doing" the action.

"while" has the same problem as "until" but "through" doesn't seem to carry the same connotation... other synonyms come to mind?

**Comment by lowne** [comment]
Yes, on second thought the purpose, behaviour and syntax are more self-explanatory if the "do" prefix is kept.

**Comment by asmagill** [comment]
went with:
	hs.timer.doUntil(predicateFn, actionFn [, interval])
	hs.timer.doWhile(predicateFn, actionFn [, interval])

	and

	hs.timer.waitUntil(predicateFn, actionFn [, interval])
	hs.timer.waitWhile(predicateFn, actionFn [, interval])

just to round things out.

Also added a method hs.timer:running() which returns true or false indicating whether or not a timer is currently running or stopped, and just as a shorthand, added hs.timer.doEvery(...) which is really just hs.timer.new(...):start()

Will create a pull request later tonight after I test them some more to make sure I have each ones logic right!

**Comment by cmsj** [comment]
Closed by /Users/cmsj/Library/Developer/Xcode/DerivedData/Hammerspoon-bcanpjyjcsxqcveiabevaoqikpya/Build/Products/Release/
